refactor :hammer: : implement client factory for DNS client [#155]

sudiptob2 opened the pull request:
## Description

closes #155 

## Changes Made

1. implemented client factory folder structure
2. implemented creation of DNSClient using factory
3. refactored `get_managed_zones` and `test_dns_policies` with DNSClient
4. updated unit test to call `get_managed_zones` and `test_dns_policies` with new method signature.

under-hill commented on src/gcp_scanner/scanner.py:
```diff
@@ -232,13 +234,15 @@ def crawl_loop(initial_sa_tuples: List[Tuple[str, Credentials, List[str]]],
-                                                                  credentials)
+        project_result['managed_zones'] = crawl.get_managed_zones(
+          project_id,
+          ClientFactory.get_client('dns').get_service(credentials),
```
Eventually I'd like to move each of these resource-specific client creation calls behind a function that takes the resource string (e.g. "dns") and creds, and leverages a dictionary mapping from resource string to service object. Then we can dynamically update the dict as we loop through the scan_config: either set the newly created object or re-use if it's already there for the resource key.

I'm open to doing this now, or revisiting after the we've implemented each of the clients. WDYT?

sudiptob2 replied:
Hi, I'm trying to gain a complete understanding of the motivation behind this approach. Is it because calling `ClientFactory.get_client('dns').get_service(credentials)` multiple times creates redundant objects in memory? If so I can think of an alternative approach also. :bulb: 

For me, it would be more convenient to incorporate the above refactoring when we implement the double-loop approach. At this moment, I'm having difficulty visualizing the overall implementation, and I think it would be easier for me to grasp if I can start by implementing the simpler parts. Once I have a better understanding of those, I believe I will be able to comprehend this approach more easily :dart:

under-hill replied:
Yes, it helps us keep only one client in memory for each resource.

That sounds perfectly reasonable, let's revisit the idea later and introduce it as an isolated PR if we decide to do it.

under-hill commented on src/gcp_scanner/client/interface_client.py:
```diff
@@ -0,0 +1,36 @@
+      credentials: An google.oauth2.credentials.Credentials object.
+
+    Returns:
+      An objet of discovery.Resource
```
nit: spelling

sudiptob2 replied:
fixed spelling

under-hill commented on src/gcp_scanner/client/dns_client.py:
```diff
@@ -0,0 +1,38 @@
+      credentials: An google.oauth2.credentials.Credentials object.
+
+    Returns:
+      An objet discovery.Resource
```
nit: spelling

under-hill commented on src/gcp_scanner/client/client_factory.py:
```diff
@@ -0,0 +1,31 @@
+
+
+class ClientFactory:
+  """Factory class for getting generating SVG."""
```
What is an SVG here?

under-hill reviewed (commented):
Some spelling / documentation nits, and discussion about resource->client map

sudiptob2 commented:
> 

Thanks a ton! I am extremely sorry that I missed some of the basic docs and spelling. will be addressing your comments one by one as soon as possible.

sudiptob2 commented on src/gcp_scanner/client/interface_client.py:
```diff
@@ -0,0 +1,36 @@
+  @staticmethod
+  @abstractmethod
+  def get_service(credentials: Credentials) -> discovery.Resource:
+    """Create a client.
```
updated.

under-hill replied:
This should be "Get a discovery resource for a client", not "Create a client", right?

sudiptob2 replied:
thanks, lets make it more verbose.

under-hill commented on src/gcp_scanner/client/interface_client.py:
```diff
@@ -0,0 +1,36 @@
+      An object of discovery.Resource
+    """
+
+    raise NotImplementedError("Child class must implement create_client")
```
do you mean "must implement get_service" here?

sudiptob2 replied:
fixed :pray:

under-hill commented on src/gcp_scanner/test_unit.py:
Let's add a test for calling ClientFactory.get_client() with a gibberish string argument, and ensuring that it logs an error and returns None as expected.

sudiptob2 replied:
added test. Also organized tests related to client factory in a Class, later on we can add at least one test case for each client factory here.

under-hill commented on src/gcp_scanner/crawl.py:
```diff
@@ -394,12 +394,12 @@ def get_bucket_names(project_name: str, credentials: Credentials,
   Args:
     project_name: A name of a project to query info about.
-    credentials: An google.oauth2.credentials.Credentials object.
+    service: A resource object for interacting with the Compute API.
```
This should say "DNS API", not "Compute API" right?

sudiptob2 replied:
yes, or can we say GCP API?

under-hill commented on src/gcp_scanner/crawl.py:
```diff
@@ -1026,18 +1023,18 @@ def list_sourcerepo(project_id: str, credentials: Credentials) -> List[Any]:
   Args:
     project_id: An id of a project to query info about.
-    credentials: An google.oauth2.credentials.Credentials object.
+    service: A resource object for interacting with the Compute API.
```
This should say "DNS API", not "Compute API" right?

sudiptob2 replied:
updated

under-hill reviewed (commented):
More comment nits, and request to add another unit test
